lib/statsd/instrument.rb

```diff
@@ -373,10 +383,17 @@ def histogram(key, value, *metric_options)
   #      end
   def distribution(key, value=nil, *metric_options, &block)
     value, metric_options = parse_options(value, metric_options)
+
     result = nil
-    value  = 1000 * StatsD::Instrument.duration { result = block.call } if block_given?
-    metric = collect_metric(:d, key, value, metric_options)
-    result = metric unless block_given?
+    start = StatsD::Instrument.current_timestamp
```

Comment:
I am not a huge fan of all these `if block_given?` even if arguably they were somewhat already present.

But this code path could only get called when a block is given (not used otherwise)

I've tried to come up with a more declarative split of the 2 flows (block vs no-block), what do you think?

```ruby
def distribution(key, value = nil, *metric_options, &block)
  value, metric_options = parse_options(value, metric_options)
  result = nil

  if block_given?
    start = StatsD::Instrument.current_timestamp
    begin
      result = block.call
    ensure
      value = 1000 * (StatsD::Instrument.current_timestamp - start)
      collect_metric(:d, key, value, metric_options)
    end
  else
    result = collect_metric(:d, key, value, metric_options)
  end

  result
end
```

Reply:
I like this!

Reply:
I tweaked it a bit more as it's easier to return early `unless block_given?`, have another look!